python/ray/data/_internal/execution/autoscaler/default_autoscaler.py

```diff
@@ -52,73 +59,60 @@ def _calculate_actor_pool_util(self, actor_pool: AutoscalingActorPool):
         else:
             return actor_pool.num_active_actors() / actor_pool.current_size()
 
-    def _actor_pool_should_scale_up(
+    def _derive_scaling_action(
         self,
         actor_pool: AutoscalingActorPool,
         op: "PhysicalOperator",
         op_state: "OpState",
-    ):
+    ) -> _AutoscalingAction:
         # Do not scale up, if the op is completed or no more inputs are coming.
-        if op.completed() or (op_state.total_enqueued_input_bundles() == 0):
-            return False
+        if op.completed() or (
+            op._inputs_complete and op_state.total_enqueued_input_bundles() == 0
+        ):
+            return _AutoscalingAction.SCALE_DOWN
+
         if actor_pool.current_size() < actor_pool.min_size():
             # Scale up, if the actor pool is below min size.
-            return True
-        elif actor_pool.current_size() >= actor_pool.max_size():
+            return _AutoscalingAction.SCALE_UP
+        elif actor_pool.current_size() > actor_pool.max_size():
             # Do not scale up, if the actor pool is already at max size.
-            return False
+            return _AutoscalingAction.SCALE_DOWN
+
         # Do not scale up, if the op does not have more resources.
         if not op_state._scheduling_status.under_resource_limits:
-            return False
+            return _AutoscalingAction.NO_OP
+
         # Do not scale up, if the op has enough free slots for the existing inputs.
         if op_state.total_enqueued_input_bundles() <= actor_pool.num_free_task_slots():
-            return False
-        # Determine whether to scale up based on the actor pool utilization.
-        util = self._calculate_actor_pool_util(actor_pool)
-        return util > self._actor_pool_scaling_up_threshold
+            return _AutoscalingAction.NO_OP
```

Comment:
I think these no-ops change the autoscaling behavior. Previously, if `util < self._actor_pool_scaling_down_threshold`, we'd scale down even if `not op_state._scheduling_status.under_resource_limits` or `op_state.total_enqueued_input_bundles() <= actor_pool.num_free_task_slots()`; now, it's a no-op.

I think this might be undesirable in some cases. For example, if an actor pool is processing data faster than it's receiving (i.e., fewer inputs than slots), then we should probably scale down rather than no-op.

Reply:
Good catch!